components/service/glean/src/main/java/mozilla/components/service/glean/net/PingUploader.kt

```diff
@@ -4,22 +4,111 @@
+            try {
+                val json = JSONObject(data)
+                val indented = json.toString(2)
+
+                logger.debug("Glean ping to URL: $path\n$indented")
+            } catch (e: JSONException) {
+                logger.debug("Exception parsing ping as JSON: $e") // $COVERAGE-IGNORE$
+            }
+        }
+    }
 
-    fun createDateHeaderValue(): String {
-        val calendar = Calendar.getInstance()
+    /**
+     * TEST-ONLY. Allows to set specific dates for testing.
+     */
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun getCalendarInstance(): Calendar { return Calendar.getInstance() }
+
+    /**
+     * Generate a date string to be used in the Date header.
+     */
+    private fun createDateHeaderValue(): String {
+        val calendar = getCalendarInstance()
         val dateFormat = SimpleDateFormat("EEE, dd MMM yyyy HH:mm:ss z", Locale.US)
         dateFormat.timeZone = TimeZone.getTimeZone("GMT")
         return dateFormat.format(calendar.time)
     }
+
+    /**
+     * Generate a list of headers to send with the request.
+     *
+     * @param config the Glean configuration object
+     * @return a [HeadersList] containing String to String [Pair] with the first
+     * entry being the header name and the second its value.
+     */
+    private fun getHeadersToSend(config: Configuration): HeadersList {
+        val headers = mutableListOf(
+            Pair("Content-Type", "application/json; charset=utf-8"),
+            Pair("User-Agent", config.userAgent),
+            Pair("Date", createDateHeaderValue()),
+            // Add headers for supporting the legacy pipeline.
+            Pair("X-Client-Type", "Glean"),
+            Pair("X-Client-Version", BuildConfig.LIBRARY_VERSION)
+        )
+
+        // If there is a pingTag set, then this header needs to be added in order to flag pings
+        // for "debug view" use.
+        config.pingTag?.let {
+            headers.add(Pair("X-Debug-ID", it))
+        }
+
+        return headers
+    }
+
+    fun doUpload(path: String, data: String, config: Configuration): Boolean {
+        logPing(path, data, config)
+
+        return upload(
+            path = path,
+            data = data,
+            headers = getHeadersToSend(config),
+            config = config)
+    }
+
+    /**
+     * Synchronously upload a ping to the server.
+     *
+     * @param path the URL path to append to the server address
+     * @param data the serialized text data to send
+     * @param config the Glean configuration object
+     *
+     * @return true if the ping was correctly dealt with (sent successfully
+     *         or faced an unrecoverable error), false if there was a recoverable
+     *         error callers can deal with.
+     */
+    abstract fun upload(path: String, data: String, headers: HeadersList, config: Configuration): Boolean
```

Comment:
API nit: As someone on the outside who would implement this, I'd expect this to be just an interface with an `upload` method signature since I wouldn't care about the other things in here. Internally Glean can still wrap this as it needs.

Reply:
I see, good point!

Reply:
@pocmo I re-flagged you for review: I'm now using Kotlin delegates to simplify this. Any chance you could give a more in-depth look to how their used in `BaseUploader` and `Glean`?